models/astrocyte_lr_1994.cpp

```diff
@@ -0,0 +1,543 @@
+  }
+
+  if ( not B_.c_ )
+  {
+    B_.c_ = gsl_odeiv_control_y_new( 1e-6, 0.0 );
+  }
+  else
+  {
+    gsl_odeiv_control_init( B_.c_, 1e-6, 0.0, 1.0, 0.0 );
+  }
+
+  if ( not B_.e_ )
+  {
+    B_.e_ = gsl_odeiv_evolve_alloc( State_::STATE_VEC_SIZE );
+  }
+  else
+  {
+    gsl_odeiv_evolve_reset( B_.e_ );
+  }
+
+  B_.sys_.function = astrocyte_lr_1994_dynamics;
+  B_.sys_.jacobian = nullptr;
+  B_.sys_.dimension = State_::STATE_VEC_SIZE;
+  B_.sys_.params = reinterpret_cast< void* >( this );
+
+  B_.I_stim_ = 0.0;
+}
+
+void
+nest::astrocyte_lr_1994::pre_run_hook()
+{
+  // ensures initialization in case mm connected after Simulate
+  B_.logger_.init();
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ *
+ * Spikes are not generated by astrocytes and the related functions
+ * are not needed. However, it is kept here to avoid possible
+ * inconsistencies with the rest of the code.
+ * TO DO: Check how to safely remove spike handling functions.
+ * ---------------------------------------------------------------- */
+
+inline void
+nest::astrocyte_lr_1994::update( Time const& origin, const long from, const long to )
+{
+  for ( long lag = from; lag < to; ++lag )
+  {
+    double t = 0.0;
+
+    // numerical integration with adaptive step size control:
+    // ------------------------------------------------------
+    // gsl_odeiv_evolve_apply performs only a single numerical
+    // integration step, starting from t and bounded by step;
+    // the while-loop ensures integration over the whole simulation
+    // step (0, step] if more than one integration step is needed due
+    // to a small integration step size;
+    // note that (t+IntegrationStep > step) leads to integration over
+    // (t, step] and afterwards setting t to step, but it does not
+    // enforce setting IntegrationStep to step-t; this is of advantage
+    // for a consistent and efficient integration across subsequent
+    // simulation intervals
+    while ( t < B_.step_ )
+    {
+      const int status = gsl_odeiv_evolve_apply( B_.e_,
+        B_.c_,
+        B_.s_,
+        &B_.sys_,             // system of ODE
+        &t,                   // from t
+        B_.step_,             // to t <= step
+        &B_.IntegrationStep_, // integration step size
+        S_.y_ );              // neuronal state
+      if ( status != GSL_SUCCESS )
+      {
+        throw GSLSolverFailure( get_name(), status );
+      }
+    }
+
+    // We limit cytosolic calcium concentration within boudnaries. Here, total
+    // volume of astrocyte consists of cytosol and endoplasmic retimulum (ER):
+    // V_tot = V_cyt + V_ER. So, ratio of cytosol vs. ER (ratio_ER_cyt) is
+    // (V_tot - V_cyt)/V_cyt. Then, cytosolic calcium concentration should not
+    // be larger than Ca_tot*(V_tot/V_cyt), which is equal to
+    // Ca_tot + Ca_tot*ratio_ER_cyt.
+    if ( S_.y_[ State_::Ca ] > P_.Ca_tot_ + P_.Ca_tot_ * P_.ratio_ER_cyt_ )
```

Comment:
simplify like this and move to `..._dynamics_` to ensure condition is fulfilled at all times.
```c++
V_.max_ca = P_.Ca_tot_ + P_.Ca_tot_ * P_.ratio_ER_cyt_;
S_.y_[ State_::Ca ] = std::max( 0, std::min( S_.y_[ State_::Ca ], V_.max_Ca ) );
```